Bug 1117583 - Make a custom form for filing intermittent bugs

KWierso opened the pull request:
Apparently I can't reopen a closed pull request if I've force-pushed to it in the time it was closed. Fun.

Anyway... This is very close to being ready for review. Still need to fix up a few things, but I'm hoping to get that done by the end of the week.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1223)
<!-- Reviewable:end -->

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

edmorley commented:
I've filed bug 1235619 for seeing if autolander can just leave the "no bug number found" message once per PR.

edmorley commented:
Just some passing comments before Cameron takes a look :-)

The API probably needs some tests.
Could you also break this into two commits - one for adding the API parts and the other for the UI work?

---

Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[treeherder/config/settings.py, line 14 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGDwWbaiO6u7jgMxc:-K6jGDwX_NNsONIlPfW3:1847998296) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/treeherder/config/settings.py#L14)):*
Could you put this next to `BZ_API_URL` and call it `BZ_API_KEY` please :-)

---

*[treeherder/webapp/api/bugzilla.py, line 20 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGTep4YziLAHR_t7E:-K6jGTep4YziLAHR_t7F:1043957832) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/treeherder/webapp/api/bugzilla.py#L20)):*
This should use `BZ_API_URL` instead of hardcoding the URL.

---

*[treeherder/webapp/api/bugzilla.py, line 22 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGZMpXxwd9fI3Ahwq:-K6jGZMpXxwd9fI3Ahwr:1815973159) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/treeherder/webapp/api/bugzilla.py#L22)):*
(Needs to be removed)

---

*[ui/index.html, line 21 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGjEwvDrEsMCc0Zmx:-K6jGjEwvDrEsMCc0Zmy:384958661) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/index.html#L21)):*
Perhaps `treeherder-bugfiler.css` or something similar, would be clearer?

---

*[ui/js/controllers/main.js, line 480 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHOvcQGMJoOW2W7bw:-K6jHOvcQGMJoOW2W7bx:-1160128187) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L480)):*
These and the other bug filer code probably belongs in it's own file, main.js is pretty long as it is hehe.

---

*[ui/js/controllers/main.js, line 528 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGuhv2sWQ3Lf_K7cF:-K6jGuhwKBfKHRBe7CYr:1630192090) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L528)):*
I'm a bit uneasy about having this massive hardcoded list of mappings. Did you speak to gps about the hg.m.o API for fetching metadata about files? (Per https://bugzilla.mozilla.org/show_bug.cgi?id=1117583#c4)

If the API is broken, we should get it fixed. If some files are missing metadata - let's add it there instead (should be fairly quick - child directories inherit their moz.build parents metadata).

---

*[ui/js/controllers/main.js, line 634 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jH_oQorjQJSi5FB01:-K6jH_oQorjQJSi5FB02:563806338) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L634)):*
This needs cleaning up

---

*[ui/js/controllers/main.js, line 781 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHyzpwkKiiMQvfcp5:-K6jHyzpwkKiiMQvfcp6:-1783528533) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L781)):*
We should probably factor out this and the other BZ API URLs (see `services/main.js:34` for another Bugzilla instance too).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-:-K6jIHYakjtA-luHmQJZ:1230131646)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 480 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHOvcQGMJoOW2W7bw:-K6kbTNHmPNELXJ9TqMA:191948527) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L480)):*
I'm trying to split this out to somewhere else, but I don't know enough Angular-fu to do it. So far I've moved fileBug() into ui/plugins/failure_summary/controller.js because that seemed like a fitting place for it (next to the file that holds the new "file bug" button). I've also moved all of IntermittentFilerCtrl into a new file, ui/js/controllers/bugfiler.js and renamed the controller to "BugFilerCtrl" for good measure. But now when I click that button, I get an error:
Error: [ng:areq] Argument 'BugFilerCtrl' is not a function, got undefined

I've pushed a followup commit with what I have. Any pointers would be helpful. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 480 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHOvcQGMJoOW2W7bw:-K6l-4BJAwd7OFargYDb:-545945500) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L480)):*
Figured it out. Have to add bugfiler.js to ui/index.html

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[treeherder/webapp/api/bugzilla.py, line 20 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGTep4YziLAHR_t7E:-K6ljGOT_cpUbVgj44vb:-747453608) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/treeherder/webapp/api/bugzilla.py#L20)):*
Done, defaults to bmo, can be set to bugzilla-dev with an env variable. Though this now switches this for all use cases; whatever was around before the bug filer existed will now also query to bugzilla-dev if you set this. Unsure if that matters.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 634 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jH_oQorjQJSi5FB01:-K6lkC5miAydvySNjR8_:207365229) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L634)):*
Yeah, this is going to be where I build up a bunch of default products to suggest when nothing better is found. Just haven't gotten to it yet. :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 781 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHyzpwkKiiMQvfcp5:-K6lkPGAjQ8p_Urr9gJk:-1679996409) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L781)):*
Partially done elsewhere, still need to get this one and any others in the tree. How do I (can I?) access the one from django settings? Or should I just duplicate that on the UI side some other way?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 528 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGuhv2sWQ3Lf_K7cF:-K6ll2FGOgl35f4goR7g:-1051051620) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L528)):*
One thing I like having this list as a starting point is that it's instantaneous. No need to hit any network resources unless you need to find a better product/component pair. If the initial load of the filing form has to wait for the hg.m.o API, it either needs to not display until that gets back or the form's going to shift around as it comes in. 

I agree that getting the metadata into hg.m.o is a good end goal, but this seemed faster and less permanent. (One blob of json vs dozens of files around the tree)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
Now that I've moved around the new controller and stuff, I'll split this into the API/UI commits tomorrow.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-:-K6lm5E2GM4iJ-mFYj4P:-445084804)*
<!-- Sent from Reviewable.io -->

KWierso commented:
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/js/controllers/main.js, line 480 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHOvcQGMJoOW2W7bw:-K6lmMgy7fualBo7XLHf:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L480)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
Okay, API and UI have been split.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-:-K6pQWgGTqkovyOVQ9-n:1332370208)*
<!-- Sent from Reviewable.io -->

camd commented:
Reviewed 5 of 9 files at r1, 1 of 8 files at r2, 1 of 8 files at r3.
Review status: 3 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[treeherder/webapp/api/bugzilla.py, line 12 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7Y9zLIAV6vlIyS4F6k:-K7Y9zLIAV6vlIyS4F6l:-1283126454) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/treeherder/webapp/api/bugzilla.py#L12)):*
can we remove this commented out line, too?

---

*[treeherder/webapp/api/bugzilla.py, line 30 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7Y9Li8hajPvAd5UQxY:-K7Y9Li9sQyzQswKCwI4:1249730955) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/treeherder/webapp/api/bugzilla.py#L30)):*
I presume these commented out lines can be removed here and below?

---

*[ui/partials/main/intermittent.html, line 52 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7YQrqtCZiQfF68URey:-K7YQrqtCZiQfF68URez:461180636) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/ui/partials/main/intermittent.html#L52)):*
if this isn't going to work on the first iteration, we should just remove the field and add it in later.

---

*[ui/css/treeherder-intermittent.css, line 9 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7YRev_wL4IbufO8Lpg:-K7YRev_wL4IbufO8Lph:-567421878) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/css/treeherder-intermittent.css#L9)):*
remove this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

camd commented:
I threw quite a bit of info at you in the second commit.  If these concepts feel pretty new, I'd be happy to chat with you on Vidyo and we could walk you through a couple.  Please ping me on irc if you have any questions.

This is going to be a really awesome feature!

KWierso commented on treeherder/webapp/api/bugzilla.py:
```diff
@@ -0,0 +1,46 @@
+            'product': params["product"],
+            'component': params["component"],
+            'summary': params["summary"],
+            'keywords': 'intermittent-failure',
```
Now that I'm thinking about it, I should probably not hardcode the intermittent keyword into this so it can be reused for other things.

KWierso commented:
*[treeherder/webapp/api/bugzilla.py, line 12 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7Y9zLIAV6vlIyS4F6k:-K7nlvlWg5Tu4hYde9oC:-2009141057) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/treeherder/webapp/api/bugzilla.py#L12)):*
Removed it since it was unused, but I thought I was needing to add it before this landed. Ed had suggested I add it at Mozlando, but I don't remember why...

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[treeherder/webapp/api/bugzilla.py, line 29 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6dpcvu8hCN6HjQOoAK-r3-29:-K7nmCBIVSg6OuGVRDFm:528726276) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/treeherder/webapp/api/bugzilla.py#L29)):*
I switched it so the UI submits the keyword and the API doesn't expect it. Haven't really tested what happens if the UI doesn't submit any keywords...

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[treeherder/webapp/api/bugzilla.py, line 30 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7Y9Li8hajPvAd5UQxY:-K7nmPvE_yjm-gX0vaQh:894935110) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/treeherder/webapp/api/bugzilla.py#L30)):*
Removed them here and on the UI side. I left comments for what remains to be added in the future. Happy to remove even those if needed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 634 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jH_oQorjQJSi5FB01:-K7nmhYObbMhvVG8AS6y:1928488961) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L634)):*
I ended up removing this for now. If we want it back later, we can add it back later.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/js/controllers/main.js, line 528 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGuhv2sWQ3Lf_K7cF:-K7nmnXHUnrPWakRw8-I:-1930258402) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L528)):*
I moved this all into values.js and switched things to use that. Seems like a better place for them to sit for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/partials/main/intermittent.html, line 52 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7YQrqtCZiQfF68URey:-K7nn16oeHMWlj4MzLhS:-379446166) ([raw file](https://github.com/mozilla/treeherder/blob/55028f6a696287585fd35764f1073eae04f542ca/ui/partials/main/intermittent.html#L52)):*
I removed all of the non-implemented parts for now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
*[ui/css/treeherder-intermittent.css, line 9 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7YRev_wL4IbufO8Lpg:-K7nn6jUmG24W87wVnb3:677251183) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/css/treeherder-intermittent.css#L9)):*
Yep.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223)*
<!-- Sent from Reviewable.io -->

KWierso commented:
Addressed most of this. Still need to AngularJSify a bunch of things, but this is most of the review comments addressed.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-:-K7nnDFrShRKIigw3r5x:-132205502)*
<!-- Sent from Reviewable.io -->

edmorley commented:
Thank you for the tweaks so far, looking forward to seeing this land :-)

---

Review status: 1 of 11 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[treeherder/config/settings.py, line 322 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7p7--XFXOkMxF7G0BC:-K7p7--XFXOkMxF7G0BD:1669735409) ([raw file](https://github.com/mozilla/treeherder/blob/12cb6af90d13c633d5fbe336e0dae50a987dea3e/treeherder/config/settings.py#L322)):*
I don't think we need this change; for testing locally override the current setting using local.py
(To avoid this getting checked in)

---

*[treeherder/config/settings.py, line 323 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7p78CeRcSc8jwPPnvm:-K7p78CeRcSc8jwPPnvn:-91790659) ([raw file](https://github.com/mozilla/treeherder/blob/12cb6af90d13c633d5fbe336e0dae50a987dea3e/treeherder/config/settings.py#L323)):*
Please drop the TREEHERDER_ prefix, we're moving away from it

---

*[ui/js/controllers/bugfiler.js, line 122 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K7p9cQwpgKxqcnSpC5-:-K7p9cQxY6FEpBYDf5Ce:988961946) ([raw file](https://github.com/mozilla/treeherder/blob/12cb6af90d13c633d5fbe336e0dae50a987dea3e/ui/js/controllers/bugfiler.js#L122)):*
Missing semicolon:
https://travis-ci.org/mozilla/treeherder/jobs/101742279

---

*[ui/js/controllers/main.js, line 528 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jGuhv2sWQ3Lf_K7cF:-K7p8du7sxx-6ZiE3AKa:813783532) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L528)):*
Dozens of json blobs around the tree isn't an issue; it's desired to have state tracked in the repo, and is an explicit end state of gps' plan etc. Please can you file the bugs to at least get the ball rolling here? It's been 3 months since https://bugzilla.mozilla.org/show_bug.cgi?id=1117583#c4 , it would be good to get movement there sooner rather than later to give gps et al time to fix it :-)

The response time from hg.m.o shouldn't be a problem - the intermittent bug filer form already has to load stuff from Treeherder's API. The layout adjusting could be handled by making the input fields large enough to handle 95% of the component names.

---

*[ui/js/controllers/main.js, line 781 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-K6jHyzpwkKiiMQvfcp5:-K7p8NmIBcHBqEK4-OQ3:1253803316) ([raw file](https://github.com/mozilla/treeherder/blob/b7eaac5c779518aa9d33ac98474f72ffe7c6c70a/ui/js/controllers/main.js#L781)):*
Duplicate on the UI side (see above comment that mentions an example to copy)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1223#-:-K7p9hNA3xbs8I-hp6vP:-773738591)*
<!-- Sent from Reviewable.io -->

KWierso commented:
This is now superseded by PR 1374.
